### PR TITLE
Fixing enum comparison at where clause to TRY_CAST

### LINKED ISSUE #3641: [R and Python] Cannot run query that has values not present in factor/Categorical column
This will throw an error:
```r
library(DBI)
library(duckdb)

df <- data.frame(
    col1 = factor(c("a", "b", "c"))
)
q <- "select * from df where col1 != ''"
conn <- dbConnect(duckdb::duckdb(), dbdir=":memory:", read_only=FALSE)
duckdb_register(conn, "df", df)
dbGetQuery(conn, q)
```
```
Error: duckdb_prepare_R: Failed to prepare query select * from df where col1 != ''
Error: Conversion Error: Could not convert string '' to UINT8
```

I guess it is trying to convert `''` to its underlying integer representation in the factor variable, but there is no mapping for it. It shouldn't require values passed in the query to already be a part of the factor encoding.

The same thing happens if I try something like this:
```sql
select col1
from df
    union all
select 'd' as col1
```
```
Error: duckdb_execute_R: Failed to run query
Error: Conversion Error: Could not convert string 'd' to UINT8
```

The problem also occurs in the Python interface:
```python
import pandas as pd
import duckdb

df = pd.DataFrame({
    "col1" : pd.Categorical(pd.Series(["a", "b", "c"]))
})
duckdb.query(
    "select * from df where col1 != ''"
).to_df()
```
```
RuntimeError: Conversion Error: Could not convert string '' to UINT8
```

### PR DESCRIPTION
**Pull request by pdet**
Fixes: #3641

### CODE REVIEW

**Comment by pdet** [comment]
@Mytherin, I'm not particularly happy with this solution, seems a bit hacky, maybe you have a suggestion?

File: src/planner/binder/expression/bind_comparison_expression.cpp [hunk 1]
```diff
@@ -125,6 +125,14 @@ BindResult ExpressionBinder::BindExpression(ComparisonExpression &expr, idx_t de
 	// add casts (if necessary)
 	left.expr = BoundCastExpression::AddCastToType(move(left.expr), input_type);
 	right.expr = BoundCastExpression::AddCastToType(move(right.expr), input_type);
+	if (where_clause) {
+		if (left.expr->type == ExpressionType::OPERATOR_CAST && left.expr->return_type.id() == LogicalTypeId::ENUM) {
```
Comment (Mytherin) [inline src/planner/binder/expression/bind_comparison_expression.cpp:129]: Is there a reason to only do this in the `WHERE` clause? Should comparisons now throw an error if I do e.g. `SELECT enum='' FROM tbl`

File: src/planner/binder/expression/bind_comparison_expression.cpp [hunk 2]
```diff
@@ -125,6 +125,14 @@ BindResult ExpressionBinder::BindExpression(ComparisonExpression &expr, idx_t de
...
+			((BoundCastExpression *)left.expr.get())->try_cast = true;
+		}
+		if (right.expr->type == ExpressionType::OPERATOR_CAST && right.expr->return_type.id() == LogicalTypeId::ENUM) {
```
Comment (Mytherin) [inline src/planner/binder/expression/bind_comparison_expression.cpp:132]: Should we have different behavior if the user inserts a cast, vs when we do an implicit cast? Currently the two are treated the same:

```sql
-- implicit cast
SELECT * FROM tbl WHERE ecol=e2col
-- explicit cast
SELECT * FROM tbl WHERE ecol::ENUM2=e2col
```

File: src/planner/binder/expression/bind_comparison_expression.cpp (hunk 1 above)
Comment (pdet) [reply src/planner/binder/expression/bind_comparison_expression.cpp:129]: Basically, we should always default to a try cast, unless is explicit?

Comment (Mytherin) [reply src/planner/binder/expression/bind_comparison_expression.cpp:129]: Perhaps the `AddCastToType` could take a `try_cast` parameter (default to false) that is set to true if any of the inputs is an enum?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**